beacon_node/beacon_chain/src/attestation_verification/batch.rs

```diff
@@ -0,0 +1,229 @@
+            if result.is_ok() {
+                num_indexed += 1;
+            } else {
+                num_failed += 1;
+            }
+            result
+        })
+        .collect::<Vec<_>>();
+
+    // May be set to `No` if batch verification succeeds.
+    let mut check_signatures = CheckAttestationSignature::Yes;
+
+    // Perform batch BLS verification, if any attestation signatures are worth checking.
+    if num_indexed > 0 {
+        let signature_setup_timer =
+            metrics::start_timer(&metrics::ATTESTATION_PROCESSING_SIGNATURE_SETUP_TIMES);
+
+        let pubkey_cache = chain
+            .validator_pubkey_cache
+            .try_read_for(VALIDATOR_PUBKEY_CACHE_LOCK_TIMEOUT)
+            .ok_or(BeaconChainError::ValidatorPubkeyCacheLockTimeout)?;
+
+        let fork = chain
+            .canonical_head
+            .try_read_for(HEAD_LOCK_TIMEOUT)
+            .ok_or(BeaconChainError::CanonicalHeadLockTimeout)
+            .map(|head| head.beacon_state.fork())?;
+
+        let mut signature_sets = Vec::with_capacity(num_indexed * 3);
+
+        // Iterate, flattening to get only the `Ok` values.
+        for indexed in indexing_results.iter().flatten() {
+            let signed_aggregate = &indexed.signed_aggregate;
+            let indexed_attestation = &indexed.indexed_attestation;
+
+            signature_sets.push(
+                signed_aggregate_selection_proof_signature_set(
+                    |validator_index| pubkey_cache.get(validator_index).map(Cow::Borrowed),
+                    signed_aggregate,
+                    &fork,
+                    chain.genesis_validators_root,
+                    &chain.spec,
+                )
+                .map_err(BeaconChainError::SignatureSetError)?,
+            );
+            signature_sets.push(
+                signed_aggregate_signature_set(
+                    |validator_index| pubkey_cache.get(validator_index).map(Cow::Borrowed),
+                    signed_aggregate,
+                    &fork,
+                    chain.genesis_validators_root,
+                    &chain.spec,
+                )
+                .map_err(BeaconChainError::SignatureSetError)?,
+            );
+            signature_sets.push(
+                indexed_attestation_signature_set_from_pubkeys(
+                    |validator_index| pubkey_cache.get(validator_index).map(Cow::Borrowed),
+                    &indexed_attestation.signature,
+                    indexed_attestation,
+                    &fork,
+                    chain.genesis_validators_root,
+                    &chain.spec,
+                )
+                .map_err(BeaconChainError::SignatureSetError)?,
+            );
+        }
+
+        metrics::stop_timer(signature_setup_timer);
+
+        let _signature_verification_timer =
+            metrics::start_timer(&metrics::ATTESTATION_PROCESSING_SIGNATURE_TIMES);
+
+        if verify_signature_sets(signature_sets.iter()) {
```

Comment:
It might be overdoing it, but we could potentially parallelise this using the `ParallelSignatureSets` type that was extracted from parallel block verification on the checkpoint sync branch:

https://github.com/sigp/lighthouse/blob/c85c8ab61c48acfecfb9fcfc604ac5465ef945ed/consensus/state_processing/src/per_block_processing/block_signature_verifier.rs#L333-L361

It would need a new home outside `per_block_processing` though

Reply:
I also considered this, but I had the feeling that we'll probably already be doing concurrent verification of batches so there wouldn't be a whole lot to gain.

Perhaps we can experiment once we get this base version out the door. I wouldn't mind experimenting more with the batch size, as well.

Reply:
Totes, sounds good